File: pipeline/presets.py
"""Hard-coded scan_replica preset for the oil-texture bump pipeline.

scan_replica mimics a real 3D scan of an oil painting: paint-density-driven
mid-gray baseline with subtle impasto ridges where the input image has
visible brush-stroke structure, plus a luminance-driven DC offset so bright
regions read as physically raised paint and dark regions as recessed.

Tune the numbers below; they take effect on the next pipeline run.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class StylePreset:
    """Locked parameter set for the scan_replica style."""

    name: str

    # Anisotropic LIC stroke field
    stroke_length: float
    stroke_thickness: float
    direction_strength: float

    # Pre-blur sigma applied to luminance before the structure tensor.
    # > 0 turns the orientation field into a HIGH-PASS structure tensor, so
    # large-scale silhouette edges (subject vs background) stop dominating
    # the LIC direction; 0 follows the global edges.
    orientation_highpass_sigma: float

    # Weight on the isotropic-noise fallback used by the LIC where coherence
    # is low. 0.0 = smooth/low-coherence regions return to flat (no noise).
    iso_weight: float

    # Power applied to the paint-density (variance-driven) thickness mask.
    # >1 emphasizes strongly-textured regions over weakly-textured ones.
    thickness_gamma: float

    # Lift the paint-density mask so even regions with raw_density == 0
    # still receive `floor` of the texture. 0 = pure variance-driven; 0.4
    # gives a uniform "canvas of paint" base + extra impasto in detail
    # regions.
    thickness_floor: float

    # Final bump-map amplitude before quantization. The composer rescales
    # the 99th-percentile of |bump| to this value.
    output_amplitude: float

    # Luminance-driven DC offset. > 0 makes bright textured regions
    # persistently raised, dark textured regions recessed; smooth regions
    # stay at mid-gray. Without this, the bump map only has zero-mean
    # ridges (AC) and looks like brushy fluff instead of real 3D-scan
    # height.
    luminance_height_bias: float

    # E1 anti-smoothing pre-compensation
    unsharp_sigma: float
    unsharp_alpha: float


PRESET = StylePreset(
    name="scan_replica",
    stroke_length=32.0,
    stroke_thickness=4.0,
    direction_strength=1.0,
    orientation_highpass_sigma=8.0,
    iso_weight=0.0,
    thickness_gamma=2.0,
    thickness_floor=0.0,
    output_amplitude=0.22,
    luminance_height_bias=0.5,
    unsharp_sigma=1.0,
    unsharp_alpha=0.5,
)

File: web/src/presets.js
// scan_replica 风格预设。数值与 pipeline/presets.py 中的 PRESET 一一对应。
//
// 调参时直接改这里：刷新页面就生效。

/**
 * @typedef {Object} StylePreset
 * @property {string} name
 * @property {number} strokeLength             - LIC 笔触长度 (px)
 * @property {number} strokeThickness          - LIC 笔触厚度 (px)
 * @property {number} directionStrength        - 0=各向同性, 1=完全 LIC
 * @property {number} orientationHighpassSigma - 结构张量前的 luminance 高通 σ
 * @property {number} isoWeight                - 低 coherence 区域的各向同性 fallback 权重
 * @property {number} thicknessGamma           - paint-density mask 的 power
 * @property {number} thicknessFloor           - paint-density mask 的下限
 * @property {number} outputAmplitude          - 99 百分位 |bump| 目标
 * @property {number} luminanceHeightBias      - 0=纯 ridge field, 0.5=明显 impasto
 * @property {number} unsharpSigma             - E1 反平滑 σ
 * @property {number} unsharpAlpha             - E1 反平滑强度
 */

/** @type {StylePreset} */
export const PRESET = Object.freeze({
    name: 'scan_replica',
    strokeLength: 32.0,
    strokeThickness: 4.0,
    directionStrength: 1.0,
    orientationHighpassSigma: 8.0,
    isoWeight: 0.0,
    thicknessGamma: 2.0,
    thicknessFloor: 0.0,
    outputAmplitude: 0.22,
    luminanceHeightBias: 0.5,
    unsharpSigma: 1.0,
    unsharpAlpha: 0.5,
});
